File: old/algo/147.kt
data class ListNode(
    var `val`: Int,
    var next: ListNode? = null
)

//class Solution {
//    fun insertionSortList(head: ListNode?): ListNode? {
//        val resultHead = ListNode(Int.MIN_VALUE)
//        var node = head
//
//        outer@ while (node != null) {
//            var resultNode: ListNode? = resultHead
//            var prevResultNode: ListNode? = null
//            while (resultNode != null) {
//                if (resultNode.`val` > node!!.`val`) {
//                    prevResultNode?.next = node
//                    val temp = node.next
//                    node.next = resultNode
//                    node = temp
//                    continue@outer
//                }
//                prevResultNode = resultNode
//                resultNode = resultNode.next
//            }
//            prevResultNode?.next = node
//            val temp = node?.next
//            node?.next = null
//            node = temp
//        }
//        return resultHead.next
//    }
//}

// 항상 처음부터 찾던 코드를 개선.

class Solution {
    fun insertionSortList(head: ListNode?): ListNode? {
        val resultHead = ListNode(Int.MIN_VALUE)
        var node = head
        var resultNode: ListNode? = null
        var prevResultNode: ListNode? = null

        outer@ while (node != null) {
            if (resultNode == null || resultNode.`val` > node.`val`) {
                resultNode = resultHead
                prevResultNode = null
            }
            while (resultNode != null) {
                if (resultNode.`val` > node!!.`val`) {
                    prevResultNode?.next = node
                    val temp = node.next
                    node.next = resultNode
                    node = temp
                    continue@outer
                }
                prevResultNode = resultNode
                resultNode = resultNode.next
            }
            prevResultNode?.next = node
            val temp = node?.next
            node?.next = null
            node = temp
        }
        return resultHead.next
    }
}

val head = ListNode(0)
var node: ListNode? = head

//val list = intArrayOf(4, 2, 1, 3)
//val list = intArrayOf(-1,5,3,4,0)
val list = intArrayOf(4, 19, 14, 5, -3, 1, 8, 5, 11, 15)
for (i in list) {
    node?.next = ListNode(i)
    node = node?.next
}

var newNode: ListNode? = Solution().insertionSortList(head.next)
while (newNode != null) {
    println(newNode?.`val`)
    newNode = newNode?.next
}

File: old/algo/147.py
from typing import Optional
import sys

class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

# 1600, 1500ms대 타임 아웃 위기.
# class Solution:
#     def insertionSortList(self, head: Optional[ListNode]) -> Optional[ListNode]:
#         current = root = ListNode(-sys.maxsize)

#         while head:
#             # and current.next.val and 조건을 넣는다면 0 떄문에 걸림. 주의
#             while current.next and current.next.val < head.val:
#                 current = current.next
#             current.next, head.next, head = head, current.next, head.next
#             current = root

#         return root.next

# 161ms
class Solution:
    def insertionSortList(self, head: Optional[ListNode]) -> Optional[ListNode]:
        current = root = ListNode(-sys.maxsize)

        while head:
            while current.next and current.next.val < head.val:
                current = current.next
            current.next, head.next, head = head, current.next, head.next

            if head and head.val < current.val:
                current = root

        return root.next


a = Solution()
head = ListNode(4)
head.next = ListNode(2)
head.next.next = ListNode(1)
head.next.next.next = ListNode(3)

result = a.insertionSortList(head)
while result:
    print(result.val)
    result = result.next

# head = [4,2,1,3]

head = ListNode(-1)
head.next = ListNode(5)
head.next.next = ListNode(3)
head.next.next.next = ListNode(4)
head.next.next.next.next = ListNode(0)

result = a.insertionSortList(head)
while result:
    print(result.val)
    result = result.next

# head = [-1,5,3,4,0]